NotReadableError: The I/O read operation failed for large files of size 4G+

youennf opened the pull request:
#### 0608392ae72e8a6c652e8cd407d0cceade20e3c2
<pre>
NotReadableError: The I/O read operation failed for large files of size 4G+
<a href="https://bugs.webkit.org/show_bug.cgi?id=272600">https://bugs.webkit.org/show_bug.cgi?id=272600</a>
<a href="https://rdar.apple.com/126863800">rdar://126863800</a>

Reviewed by Alex Christensen.

FileReaderLoader is limiting the size of data that can be read.
This restriction is not needed when reading data via a ReadableStream.
We stop allocating a buffer in FileReaderLoader for ReadType::ReadAsBinaryChunks as this is unnecessary.
Instead, we let the client store the buffers.

Manually tested for big files.

* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::didReceiveResponse):
* Source/WebCore/fileapi/FileReaderLoader.h:

Canonical link: <a href="https://commits.webkit.org/278457@main">https://commits.webkit.org/278457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc1d051693bdcf7271cea82e2b38f635aa80d143

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53853 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1284 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36146 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/935 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41250 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52693 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27544 "") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43573 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22369 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24935 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/831 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9034 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/894 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55442 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25695 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/798 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48657 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26953 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47713 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11086 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27819 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26685 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/fileapi/FileReaderLoader.cpp:
```diff
@@ -155,18 +158,25 @@ void FileReaderLoader::didReceiveResponse(ResourceLoaderIdentifier, const Resour
     // FIXME: Support reading more than the current size limit of ArrayBuffer.
```
Can we remove this then?

youennf replied:
Not really since this is still an issue for trying to read blobs as a string or a full array buffer

achristensen07 commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -321,40 +321,95 @@ ExceptionOr<Ref<ReadableStream>> Blob::stream()
         UniqueRef<FileReaderLoader> m_loader;
+        Deque<Ref<FragmentedSharedBuffer>> m_queue;
         bool m_isStarted { false };
+        bool m_isCancelled { false };
```
Could we make a State enum instead of all these bools?

youennf replied:
Done in https://github.com/WebKit/WebKit/pull/28223

achristensen07 reviewed (approved):
This is complicated enough that I think it needs some testing.  Maybe an API test that makes a mock object at the FileReaderLoader/ThreadableLoader level and feeds in some interesting calls with synthetic sizes, errors, callbacks after cancel, etc.

youennf committed:
NotReadableError: The I/O read operation failed for large files of size 4G+
https://bugs.webkit.org/show_bug.cgi?id=272600
rdar://126863800

Reviewed by Alex Christensen.

FileReaderLoader is limiting the size of data that can be read.
This restriction is not needed when reading data via a ReadableStream.
We stop allocating a buffer in FileReaderLoader for ReadType::ReadAsBinaryChunks as this is unnecessary.
Instead, we let the client store the buffers.

Manually tested for big files.

* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::didReceiveResponse):
* Source/WebCore/fileapi/FileReaderLoader.h:

Canonical link: https://commits.webkit.org/278457@main
